src/rdkafka_msg.c

```diff
@@ -432,11 +434,23 @@ int rd_kafka_produce_batch (rd_kafka_topic_t *app_rkt, int32_t partition,
                 /* Two cases here:
                  *  partition==UA:     run the partitioner (slow)
                  *  fixed partition:   simply concatenate the queue to partit */
-                if (partition == RD_KAFKA_PARTITION_UA) {
-                        /* Partition the message */
-                        rkmessages[i].err =
-                                rd_kafka_msg_partitioner(rkt, rkm,
-                                                         0/*already locked*/);
+                if (multiple_partitions) {
+                        if (partition == RD_KAFKA_PARTITION_UA) {
+                            /* Partition the message */
+                            rkmessages[i].err =
+                            rd_kafka_msg_partitioner(rkt, rkm,
+                                                     0/*already locked*/);
+                        } else {
+                            if((i == 0) || ((i + 1 < message_cnt) && (rkmessages[i - 1].partition != rkmessages[i + 1].partition)))
+                            {
+                                    if (i != 0)
+                                            rd_kafka_toppar_destroy(s_rktp);
+                                    s_rktp = rd_kafka_toppar_get_avail(rkt, rkmessages[i].partition,
+                                                                       1/*ua on miss*/, &all_err);
+                            }
+                            rktp = rd_kafka_toppar_s2i(s_rktp);
```

Comment:
@edenhill,
would it be better of o entered the rktp assignment in this line inside the if block?

Reply:
toppar_s2i() is a no-op (unless ENABLE_SHAREDPTR_DEBUG is enabled) so I think it might be better to instead move the s2i call to:
`                            rd_kafka_toppar_enq_msg(rd_kafka_toppar_s2i(s_rktp), rkm);`